### PR TITLE
Fixes for TestAlgorithm

### PR DESCRIPTION
**Pull request by adayton1**
- Uses care::host_device_ptr without registered pointers and frees them
- Fixes invalid reads/writes due to incorrect array lengths
- Fixes incorrect expected values
- All GPU tests in TestAlgorithm now pass

### CODE REVIEW

File: test/TestAlgorithm.cpp [hunk 1]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
    TEST(X, gpu_test_##Y) { gpu_test_##X##Y(); } \
    static void gpu_test_##X##Y()
 
-GPU_TEST(algorithm, fill_empty) {
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:206]: These are intentionally removed as they do not test anything different than the host versions.

File: test/TestAlgorithm.cpp [hunk 2]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
-
-  care::host_device_ptr<int> a(temp, 7, "arrseven");
-  care::host_device_ptr<int> subset1(sub1, 1, "sub1");
-  care::host_device_ptr<int> subset3(sub3, 1, "sub3");
-  care::host_device_ptr<int> subsetrev(rev, 7, "rev");
+  const int size1 = 1;
+  care::host_device_ptr<int> subset1(size1, "sub1");
+
+  const int size3 = 3;
+  care::host_device_ptr<int> subset3(size3, "sub3");
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:824]: The size was incorrect previously.

File: test/TestAlgorithm.cpp [hunk 3]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
-  care::host_device_ptr<double> threshold3(thresh3, 3, "thresh3");
-  care::host_device_ptr<double> threshold1(thresh1, 1, "thresh1");
-
-  double cutoff = -10;
+  const int size1 = 1;
+  care::host_device_ptr<int> subset1(size1, "sub1");
+  care::host_device_ptr<double> threshold1(size1, "thresh1");
+
+  const int size3 = 3;
+  care::host_device_ptr<int> subset3(size3, "sub3");
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:884]: The size was incorrect previously.

File: test/TestAlgorithm.cpp [hunk 4]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
-
-  care::host_device_ptr<int> a(temp, 7, "arrseven");
-  care::host_device_ptr<int> subset1(sub1, 1, "sub1");
-  care::host_device_ptr<int> subset3(sub3, 1, "sub3");
-  care::host_device_ptr<int> subsetrev(rev, 7, "rev");
+  const int size1 = 1;
+  care::host_device_ptr<int> subset1(size1, "sub1");
+
+  const int size3 = 3;
+  care::host_device_ptr<int> subset3(size3, "sub3");
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:987]: The size was incorrect previously.

File: test/TestAlgorithm.cpp [hunk 5]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
-  care::host_device_ptr<double> threshold3(thresh3, 3, "thresh3");
-  care::host_device_ptr<double> threshold1(thresh1, 1, "thresh1");
-
-  double cutoff = -10;
+  const int size1 = 1;
+  care::host_device_ptr<int> subset1(size1, "sub1");
+  care::host_device_ptr<double> threshold1(size1, "thresh1");
+
+  const int size3 = 3;
+  care::host_device_ptr<int> subset3(size3, "sub3");
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:1046]: The size was incorrect previously.

File: test/TestAlgorithm.cpp [hunk 6]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
+  result = care::PickAndPerformFindMaxIndex<int>(a, mask, subsetrev, size7, threshold7, cutoff, &threshIdx);
   EXPECT_EQ(result, 6); // this is the index in the original array
 
   // len 3 subset
   cutoff = 5.0;
-  result = care::PickAndPerformFindMaxIndex<int>(a, mask, subset3, 3, threshold3, cutoff, threshIdx);
-  EXPECT_EQ(result, 0);
-  EXPECT_EQ(threshIdx[0], 1);
+  result = care::PickAndPerformFindMaxIndex<int>(a, mask, subset3, size3, threshold3, cutoff, &threshIdx);
+  EXPECT_EQ(result, 6);
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:1233]: These expected values were incorrect previously.

File: test/TestAlgorithm.cpp [hunk 7]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
+  EXPECT_EQ(threshIdx, 4); // this is the index in the subset
 
   // change the cutoff
   // CURRENTLY TEST GIVES DIFFERENT RESULTS WHEN COMPILED RELEASE VS DEBUG!!!
   cutoff = 5.0;
-  result = care::PickAndPerformFindMinIndex<int>(a, nullptr, subsetrev, 7, threshold7, cutoff, threshIdx);
-  EXPECT_EQ(result, -1); // this is the index in the original array
-  EXPECT_EQ(threshIdx[0], -1);
+  result = care::PickAndPerformFindMinIndex<int>(a, nullptr, subsetrev, size7, threshold7, cutoff, &threshIdx);
+  EXPECT_EQ(result, 2); // this is the index in the original array
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:1346]: These values were incorrect previously.

File: test/TestAlgorithm.cpp [hunk 8]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
-  care::host_device_ptr<double> threshold3(thresh3, 3, "thresh3");
-  care::host_device_ptr<double> threshold1(thresh1, 1, "thresh1");
-
-  double cutoff = -10;
+  const int size1 = 1;
+  care::host_device_ptr<int> subset1(size1, "sub1");
+  care::host_device_ptr<double> threshold1(size1, "thresh1");
+
+  const int size3 = 3;
+  care::host_device_ptr<int> subset3(size3, "sub3");
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:1148]: The size was incorrect previously.

File: test/TestAlgorithm.cpp [hunk 9]
```diff
@@ -203,768 +204,1239 @@ TEST(algorithm, intersectarrays) {
...
```
Comment (adayton1) [inline test/TestAlgorithm.cpp:1265]: The size was incorrect previously.

**Comment by adayton1** [comment]
I commented on all the bug fixes. The rest of the changes have to do with using host_device_ptrs as they are generally meant to be used, cleaning up that memory, and some general clean up.

**Review by dtaller — APPROVED** [review]
I'll glad that you caught the wrong size issue that crept up in a bunch of places, thanks!

Looks good to me.

**Review by liu15 — APPROVED** [review]

**Pull request merged**